### PR TITLE
Feat/nonreplayability

### PR DESCRIPTION
**Pull request by 0xpApaSmURf**
* Adds a new mapping that tracks when an EAT gets used
* Adds a require statement in the verify function that checks that an EAT has not been used
* Updated tests to add case for re-using EATs

### CODE REVIEW

File: test/AccessTokenVerifier/AccessTokenVerifier.test.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ describe("AccessTokenVerifier", function () {
     };
 
     this.value = {
-      expiry: Math.floor(new Date().getTime() / 1000) + 10,
+      expiry: Math.floor(new Date().getTime() / 1000) + 1000,
```
Comment (ra-phael) [inline test/AccessTokenVerifier/AccessTokenVerifier.test.ts:37]: Why this change?

File: contracts/AccessTokenConsumer.sol [hunk 2]
```diff
@@ -6,6 +6,7 @@ import "./IAccessTokenVerifier.sol";
 
 contract AccessTokenConsumer {
```
Comment (ra-phael) [inline contracts/AccessTokenConsumer.sol:7]: There's a lingering console.sol above

File: test/helpers/utils.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { splitSignature } from "@ethersproject/bytes";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { BigNumber, ethers } from "ethers";
+import { packParameters, signAccessToken } from "../../src/utils";
+
+export const generateEAT = async (
+  signer: SignerWithAddress,
+  domain: any,
```
Comment (ra-phael) [inline test/helpers/utils.ts:8]: Can we type that?

File: test/AccessTokenConsumer/AccessTokenConsumer.behaviour.ts [hunk 4]
```diff
@@ -1,8 +1,8 @@
 import { ethers, waffle } from "hardhat";
```
Comment (ra-phael) [inline test/AccessTokenConsumer/AccessTokenConsumer.behaviour.ts:1]: This file is quite big! Since there's a lot of tests repeated, just with different function signatures I feel we could create some abstraction to shrink the size of this file. Let's create an issue?

**Review by ra-phael — COMMENTED** [review]
NIce!
We are enforcing non-replayability at the AccessTokenConsumer level. Not necessarily suggesting we should do this but what are your thoughts on enforcing it in `AccessTokenVerifier` in a world where potentially multiple consumers point to the same verifier?

File: contracts/AccessTokenConsumer.sol [hunk 5]
```diff
@@ -66,4 +70,23 @@ contract AccessTokenConsumer {
...
+        bytes32 accessTokenHash = keccak256(abi.encodePacked(v, r, s));
+        return _accessTokenUsed[accessTokenHash];
+    }
+
+    function consumeAccessToken(
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) private {
+        bytes32 accessTokenHash = keccak256(abi.encodePacked(v, r, s));
```
[the remark this answers is not in the document]
Comment (0xpApaSmURf) [reply contracts/AccessTokenConsumer.sol:88]: Yes, we certainly should!!!! I can't fathom why I would have left this out...

File: test/AccessTokenVerifier/AccessTokenVerifier.test.ts (hunk 1 above)
Comment (0xpApaSmURf) [reply test/AccessTokenVerifier/AccessTokenVerifier.test.ts:37]: Tests fail because sometimes it takes more than 10 seconds to reach that part of the code, by which time the EAT has already expired xD

File: test/helpers/utils.ts (hunk 3 above)
Comment (0xpApaSmURf) [reply test/helpers/utils.ts:8]: Yes we can 👍

File: contracts/AccessTokenConsumer.sol [hunk 6]
```diff
@@ -6,6 +6,7 @@ import "./IAccessTokenVerifier.sol";
 
 contract AccessTokenConsumer {
     IAccessTokenVerifier private _verifier;
+    mapping(bytes32 => bool) private _accessTokenUsed;
```
[the remark this answers is not in the document]
Comment (0xpApaSmURf) [reply contracts/AccessTokenConsumer.sol:9]: > Library for managing uint256 to bool mapping in a compact and efficient way, providing the keys are sequential.

Unfortunately our keys won't be sequential so maybe this wouldn't work?

File: test/AccessTokenConsumer/AccessTokenConsumer.behaviour.ts (hunk 4 above)
Comment (0xpApaSmURf) [reply test/AccessTokenConsumer/AccessTokenConsumer.behaviour.ts:1]: Sure :) Issue would be great

Comment (0xpApaSmURf) [reply test/AccessTokenConsumer/AccessTokenConsumer.behaviour.ts:1]: Nevermind, already solved ;)

File: test/AccessTokenVerifier/AccessTokenVerifier.test.ts (hunk 1 above)
Comment (ra-phael) [reply test/AccessTokenVerifier/AccessTokenVerifier.test.ts:37]: Ooh

**Comment by 0xpApaSmURf** [comment]
Regarding enforcement at the verifier level I have a few thoughts:

* Originally it was designed so that the Verifier _only_ has logic that knows how to verify EAT formatted eip712 signed objects against a specific signer.
* Keeping it isolated allows applications to decide whether or not they want to use single-use or multiple-use EATs (potentially we can build this as a feature to toggle whether an EAT is single-use or not)
* Since EATs can only be consumed by a consumed by an application contract anyway, it makes sense that it's also the consumer tracking it as opposed to having a globally shared state for this.
* Regarding multiple consumers pointing to the same verifier, I think that's already feasible if these consumers also want to rely on the same issuers. This does mean that if multiple consumer want to rely on different issuers, a new contract has to be deployed. Maybe there's a world where the Verifying logic is separated from the KeyInfrastructure where each consumer may be able to point to a different trust architecture whilst still re-using the same Verifying logic (reconstructing EAT, EIP712 domain, hashing, signature check etc.).

**Review by ra-phael — COMMENTED** [review]
Awesome refactoring of that test file!

File: test/helpers/utils.ts (hunk 3 above)
Comment (0xpApaSmURf) [reply test/helpers/utils.ts:23]: I think I misunderstood what you meant. I thought you meant the `generateEAT` function wasn't explicitly typing the return type. What did you mean?

Comment (ra-phael) [reply test/helpers/utils.ts:23]: Ah I was referring to `domain: any` not the return type of generateEAT

**Review by ra-phael — APPROVED** [review]
